src/commons/api/rrdp.rs

```diff
@@ -457,29 +495,60 @@ impl SnapshotData {
         SnapshotData::new(random, current_objects)
     }
 
-    pub fn unpack(self) -> CurrentObjects {
+    pub fn unpack(self) -> HashMap<PublisherHandle, CurrentObjects> {
         self.current_objects
     }
 
-    pub fn elements(&self) -> Vec<&PublishElement> {
-        self.current_objects.elements()
+    // Get the approximate size for all current objects held by all publishers.
+    pub fn size_approx(&self) -> usize {
+        self.current_objects
+            .values()
+            .fold(0, |tot, objects| tot + objects.size_approx())
     }
 
-    /// Creates a new snapshot with the delta applied. This assumes
-    /// that the delta had been checked before. This should not be
-    /// any issue as deltas are verified when they are submitted.
-    pub fn with_delta(&self, random: RrdpFileRandom, elements: DeltaElements) -> SnapshotData {
-        let mut current_objects = self.current_objects.clone();
-        current_objects.apply_delta(elements);
+    pub fn current_objects_for(&self, publisher: &PublisherHandle) -> Option<&CurrentObjects> {
+        self.current_objects.get(publisher)
+    }
 
-        SnapshotData::new(random, current_objects)
+    pub fn publishers_current_objects(&self) -> &HashMap<PublisherHandle, CurrentObjects> {
+        &self.current_objects
+    }
+
+    pub fn set_random(&mut self, random: RrdpFileRandom) {
+        self.random = random;
     }
 
-    pub fn size(&self) -> usize {
+    /// Applies the delta for a publisher to this snapshot.
+    ///
+    /// This assumes that the delta had been checked before. This should not be
+    /// any issue as deltas are verified when they are submitted.
+    pub fn apply_delta(&mut self, publisher: &PublisherHandle, delta: DeltaElements) {
+        if let Some(objects) = self.current_objects.get_mut(publisher) {
+            objects.apply_delta(delta);
+            if objects.is_empty() {
+                self.current_objects.remove(publisher);
+            }
+        } else {
+            let mut objects = CurrentObjects::default();
+            objects.apply_delta(delta); // Can only happen for new publishers
```

Comment:
What can only happen for new publishers?

Reply:
pushed clarification